src/hooks/useBreakpoint.tsx

```diff
@@ -0,0 +1,37 @@
+import { useEffect, useState } from "react";
+import { BreakpointKey, Breakpoints } from "../Css";
+
+/**
+ * A React hook to return a record of responsive breakpoints that updates on resize.
+ *
+ * @example
+ * const { breakpoints } = useBreakpoint();
+ * if(breakpoints.mdAndDown) {...do something cool}
+ */
+
+type BreakpointsType = Partial<Record<BreakpointKey, boolean>>;
+
+function matchMediaBreakpoints(): BreakpointsType {
+  let bps = {} as BreakpointsType;
+  Object.entries(Breakpoints).forEach(([name, bp]) => {
+    bps[name as keyof BreakpointsType] = window.matchMedia(bp.replace("@media ", "")).matches;
+  });
+  return bps;
+}
+
+export function useBreakpoint(): { breakpoints: BreakpointsType } {
+  const [breakpoints, setBreakpoints] = useState(matchMediaBreakpoints());
+
+  useEffect(() => {
+    function handleResize() {
+      // Set breakpoints on resize
+      setBreakpoints(matchMediaBreakpoints());
+    }
+    window.addEventListener("resize", handleResize);
+    handleResize();
+    // Remove event listener on cleanup
+    return () => window.removeEventListener("resize", handleResize);
+  }, []);
+
+  return { breakpoints };
```

Comment:
**thought:** Every time we return breakpoints, then it is a new object, even if the key/value pairs didn't change. Can we ensure that the hook _only_ triggers a re-render of the calling components when there is a change to the matched breakpoints?

Reply:
Done!

Reply:
Afaiu the code as-is wouldn't trigger re-renders, it would just potentially break useMemos, i.e.:

```
const { breakpoints } = useBreakpoints();
const somethingExpensive = useMemo(() => ..., [breakpoints]);
```

...

That said, I wonder if we could just remove the hash and do:

```
const bp = useBreakpoints();
```

Granted, destructuring is super-common for hooks, but seems like we don't really need it here, :shrug: . Defer to @crshannon .

Reply:
Why wouldn't it cause the calling component to re-render? When the hook's value changes it is causing the component to re-render as far as I can tell.

Reply:
I am somewhat impartial here but like the de-structuring do we don't need `bp.`  - @bdow do you have any strong opinions on this as the blueprint guru?

Reply:
@crshannon AFAIU there is not an "extra `bp.`" b/c it just changes from:

```
const { breakpoints } = useBreakpoints();
console.log(breakpoints.smAndMd);
```

To:

```
const breakpoints = useBreakpoints();
console.log(breakpoints.smAndMd);
```

So I happened to use `bp` as a cute/shorter variable name for the non-destructured example, but they really are the same.

If anything the 2nd "not destructured" might actually be "more destructure" friendly b/c then we really could remove the `bp.` or `breakpoints.` by doing:

```
const { mdAndSm } = useBreakpoints();
if (mdAndSm) { ... }
```

Reply:
I like the idea of not having to immediately destructure, and doing `const breakpoints = useBreakpoints()` or being able to destructure for the specific breakpoint you care about... `const { mdAndDown } = useBreakpoints();`

Reply:
> Why wouldn't it cause the calling component to re-render? When the hook's value
> changes it is causing the component to re-render as far as I can tell.

Re-renders only happen when `setState`s are called, so merely asking a hook for a value, and getting back a "new" / non-stable return value, does not cause inherently cause _another_ re-render to happen.

If it did, this code as-is would infinite loop/constantly re-render, right?

If I'm following.

Reply:
Ah gotcha, gotcha, I took the "return value is new" too literally :-D, and yep, we're on the same page that deep-equaling old-vs-new `BreakpointsType` before calling `set` is a good idea.

Reply:
note: issue with `const { mdAndDown } = useBreakpoints();` is this now conflicts with the exported css breakpoints if they are being used in the same file.

`import {  mdAndDown } from "@homebound/beam";`

Reply:
@crshannon ah yeah, that's true, destructuring out literally `mdAndDown` may not be a good idea.

Reply:
just updated so we can now use `const breakpoints = useBreakpoints();` (or `const bps = useBreakpoints();` if you fancy) and implemented a `isEqual` check.